lib/useOnyx.ts

```diff
@@ -0,0 +1,144 @@
+    /**
+     * Determines if this key in this subscription is safe to be evicted.
+     */
+    canEvict?: boolean;
+
+    /**
+     * If set to false, then no data will be prefilled into the component.
+     */
+    initWithStoredValues?: boolean;
+
+    /**
+     * TODO: Check if we still need this flag and associated logic.
+     */
+    allowStaleData?: boolean;
+
+    /**
+     * Sets an initial value to be returned by the hook during the first render.
+     */
+    initialValue?: OnyxValue<TKey>;
+};
+
+function useOnyx<TKey extends OnyxKey>(key: TKey, options?: UseOnyxOptions<TKey>): OnyxValue<TKey> {
+    const [value, setValue] = useState<OnyxValue<TKey>>(options?.initialValue ?? (null as OnyxValue<TKey>));
+
+    const connectionIDRef = useRef<number | null>(null);
+
+    useEffect(() => {
+        connectionIDRef.current = Onyx.connect({
+            key: key as CollectionKeyBase,
+            callback: (val: unknown) => {
+                setValue(val as OnyxValue<TKey>);
+            },
+            initWithStoredValues: options?.initWithStoredValues,
+            waitForCollectionCallback: Onyx.isCollectionKey(key),
+        });
+
+        return () => {
+            if (!connectionIDRef.current) {
+                return;
+            }
+
+            Onyx.disconnect(connectionIDRef.current);
+        };
+    }, [key, options?.initWithStoredValues]);
+
+    /**
+     * Mimics withOnyx's checkEvictableKeys() behavior.
+     */
+    useEffect(() => {
+        if (options?.canEvict === undefined || !connectionIDRef.current) {
+            return;
+        }
+
+        if (!Onyx.isSafeEvictionKey(key)) {
+            throw new Error(`canEvict can't be used on key '${key}'. This key must explicitly be flagged as safe for removal by adding it to Onyx.init({safeEvictionKeys: []}).`);
+        }
+
+        if (options.canEvict) {
+            Onyx.removeFromEvictionBlockList(key, connectionIDRef.current);
+        } else {
+            Onyx.addToEvictionBlockList(key, connectionIDRef.current);
+        }
+    }, [key, options?.canEvict]);
+
+    return value;
+}
+
+function useOnyxWithSyncExternalStore<TKey extends OnyxKey>(key: TKey, options?: UseOnyxOptions<TKey>): OnyxValue<TKey> {
+    const connectionIDRef = useRef<number | null>(null);
+    const previousKey = usePrevious(key);
+
+    /**
+     * According to React docs, `getSnapshot` is a function that returns a snapshot of the data in the store that’s needed by the component.
+     * **While the store has not changed, repeated calls to getSnapshot must return the same value.**
+     * If the store changes and the returned value is different (as compared by Object.is), React re-renders the component.
+     *
+     * When the `key` is changed (e.g. to get a different record from a collection) and it's not yet in the cache,
+     * we return the value from the previous key to avoid briefly returning a `null` value to the component, thus avoiding a useless re-render.
```

Comment:
I raised a concern in a DM that this may led to some unexpected bugs in NewDot. Fabio's argument was that this is to omit unnecessary re-renders, but with a status property when a key is changed it will be re-rendered either way. 

I'm not convinced that this logic is needed.

Reply:
I've removed after my last changes (wasn't being able to reproduce it anymore)